std/stdio.d

```diff
@@ -1522,6 +1524,139 @@ for every line.
+            }
+
+            return buildPath(dir, name);
+        }
+
+        // Once should be enough given the randomness, but we'll give it three
+        // times to be safe.
+        foreach(i; 0 .. 3)
+        {
+            auto filename = genTempName(prefix, suffix, dir);
+
+            // O_EXCL does not support symbolic links.
+            if(dirName(filename).isSymlink)
+                filename = buildPath(absolutePath(dirName(filename), baseName(filename)));
+
+            import std.internal.cstring : tempCString;
+            version(Posix)
+            {
+                import core.sys.posix.sys.stat, core.sys.posix.fcntl;
+                enum flags = S_IRUSR | S_IWUSR;
+                auto fd = open(tempCString(filename), O_RDWR | O_CREAT | O_EXCL, flags);
+            }
+            else version(Windows)
+            {
+                enum flags = _S_IREAD | _S_IWRITE;
+                auto fd = _wsopen(tempCString!wchar(filename), O_RDWR | O_CREAT | O_EXCL, _SH_DENYNO, flags);
+            }
+            else static assert("Unsupported OS");
+
+            if(fd == -1)
+                continue;
+
+            File retval;
+            retval.fdopen(fd, "w+b", filename);
+            return retval;
+        }
+
+        throw new ErrnoException("Failed to create a temporary file.");
+    }
+
+    unittest
+    {
+        import std.algorithm, std.conv, std.file, std.path, std.process;
+
+        {
+            string name;
+            scope(exit) if(name.exists) std.file.remove(name);
+            {
+                auto file = File.tempFile();
+                name = file.name;
+                assert(!name.empty);
+                assert(buildNormalizedPath(name.dirName) == buildNormalizedPath(tempDir()));
+                assert(!name.baseName.empty);
+                assert(name.exists);
+                assert(name.isFile);
+                file.write("hello world");
+            }
+            assert(name.exists);
+            assert(name.isFile);
+            assert(readText(name) == "hello world");
+        }
+
+        auto dir = buildPath(tempDir(), "test_sub_dir_" ~ to!string(thisProcessID));
+        mkdir(dir);
+        scope(exit) if(dir.exists) rmdirRecurse(dir);
+        string name;
+        {
+            auto file = File.tempFile("foobar", ".baz", dir);
+            name = file.name;
+            assert(!name.empty);
+            assert(name.dirName == dir);
+            assert(buildNormalizedPath(name.dirName) == buildNormalizedPath(dir));
+            assert(!name.baseName.empty);
+            assert(name.baseName.startsWith("foobar"));
+            assert(name.extension == ".baz");
+            assert(name.length > "foobar".length + ".baz".length);
+            assert(name.exists);
+            assert(name.isFile);
+            file.write("D for the win!");
+        }
+        assert(name.exists);
+        assert(name.isFile);
+        assert(readText(name) == "D for the win!");
+    }
```

Comment:
All of those assertions are testing the state of a single call to tempFile to make sure that it did what it was supposed to. Splitting them up across multiple calls to tempFile would just make the tests more expensive - especially when you consider that each call to tempFile creates a file, and then that file gets deleted when the test is over. The assertions are cheap in comparison. I really don't see a problem with the assertions.

As for have a documented unit test, what would be the point? The function is very straightforward. What would an example even show that isn't obvious? And I see no point in making the tests slower by doing yet more file operations just so that we can show the function being called in the documentation - especially when there isn't much to show.